File: javaScript/palindromePartitioning.js
/* 
131. Palindrome Partitioning   My Submissions QuestionEditorial Solution
Total Accepted: 64068 Total Submissions: 232120 Difficulty: Medium
Given a string s, partition s such that every substring of the partition is a palindrome.

Return all possible palindrome partitioning of s.

For example, given s = "aab",
Return

  [
    ["aa","b"],
    ["a","a","b"]
  ]
 */
 /**
 * @param {string} s
 * @return {string[][]}
 */
var partition = function(s) {
	//corner case
	if (s === null || s === undefined) {
		return [];
	}

	if (s.length < 2) {
		return [s];
	}
    
    //initialize
    var results = [];
    var path = [];

    var isPalindrome = function(s) {
    	var left = 0;
    	var right = s.length - 1;

    	while (left < right) {
    		if (s.charAt(left) !== s.charAt(right)) {
    			return false;
    		}
    		left++;
    		right--;
    	}
    	return true;
    }

    var dfsHelper = function(s, results, path, start){
    	//base case
    	if (start === s.length) {
    		results.push(path.slice(0));
    		return;
    	}

    	//recursive case
    	for (var i = start; i < s.length; i++) {
    		var tmp = s.substring(start, i + 1);
    		if (!isPalindrome(tmp)) {
    			continue;
    		}
    		path.push(tmp);
    		arguments.callee(s, results, path, i+1);
    		path.pop(path.length - 1);
    	}
    }

    dfsHelper(s, results, path, 0);
    console.log(results);
    return results;
};

partition('ab');

File: java/PalindromePartitioning.java
/**
 * LeetCode 131 https://leetcode.com/problems/palindrome-partitioning/
 */

class Solution {
    List<List<String>> res = new LinkedList<>();
    LinkedList<String> path = new LinkedList<>();
    
    public List<List<String>> partition(String s) {
        backtrack(s, 0);
        return res;
    }
    
    void backtrack (String s, int startIndex) {
        // base case
        if (startIndex == s.length()) {
            res.add(new LinkedList(path));
            return;
        }
        
        for (int i = startIndex; i < s.length(); i++) {
            if(isPalindrome(s, startIndex, i)) {
                String str = s.substring(startIndex, i + 1);
                path.add(str);
            } else {
                continue;
            }
            
            backtrack(s, i+1);
            path.removeLast();
        }
    }
    
    boolean isPalindrome(String s, int start, int end) {
        while (start <= end) {
            if (s.charAt(start) != s.charAt(end)) {
                return false;
            }
            start++;
            end--;
        }
        return true;
    }
}